In [ ]:
# MODEL EVALUATION

from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
import pandas as pd

# Predictions
knn_pred = knn.predict(X_test_scaled)
svm_pred = svm.predict(X_test_scaled)
rf_pred = rfc.predict(X_test)


# Confusion Matrices

cm_knn = confusion_matrix(y_test, knn_pred)
cm_svm = confusion_matrix(y_test, svm_pred)
cm_rf = confusion_matrix(y_test, rf_pred)

print("KNN Confusion Matrix:\n", cm_knn)
print("\nSVM Confusion Matrix:\n", cm_svm)
print("\nRandom Forest Confusion Matrix:\n", cm_rf)


# Classification Reports

print("\nKNN Classification Report")
print(classification_report(y_test, knn_pred))

print("\nSVM Classification Report")
print(classification_report(y_test, svm_pred))

print("\nRandom Forest Classification Report")
print(classification_report(y_test, rf_pred))


# Accuracy

acc_knn = accuracy_score(y_test, knn_pred)
acc_svm = accuracy_score(y_test, svm_pred)
acc_rf = accuracy_score(y_test, rf_pred)


# False Positive Rate

def calculate_fpr(cm):
    tn, fp, fn, tp = cm.ravel()
    return fp / (fp + tn)

fpr_knn = calculate_fpr(cm_knn)
fpr_svm = calculate_fpr(cm_svm)
fpr_rf = calculate_fpr(cm_rf)

print("\nFalse Positive Rates")
print("KNN:", fpr_knn)
print("Random Forest:", fpr_rf)
print("SVM:", fpr_svm)


# Model Comparison Table

comparison = pd.DataFrame({
    "Model": ["KNN", "Random Forest", "SVM"],
    "Test Accuracy": [acc_knn, acc_rf, acc_svm],
    "False Positive Rate": [fpr_knn, fpr_rf, fpr_svm]
})

print("\nModel Comparison Table")
print(comparison)

KNN Confusion Matrix:
 [[198  42]
 [ 53 112]]

SVM Confusion Matrix:
 [[218  22]
 [ 29 136]]

Random Forest Confusion Matrix:
 [[212  28]
 [ 40 125]]

KNN Classification Report
              precision    recall  f1-score   support

           0       0.79      0.82      0.81       240
           1       0.73      0.68      0.70       165

    accuracy                           0.77       405
   macro avg       0.76      0.75      0.75       405
weighted avg       0.76      0.77      0.76       405


SVM Classification Report
              precision    recall  f1-score   support

           0       0.88      0.91      0.90       240
           1       0.86      0.82      0.84       165

    accuracy                           0.87       405
   macro avg       0.87      0.87      0.87       405
weighted avg       0.87      0.87      0.87       405


Random Forest Classification Report
              precision    recall  f1-score   support

           0       0.84      0.88      0.86       

Model Recommendation


After evaluating the performance of the three models — KNN, Random Forest, and Support Vector Machine (SVM) — using confusion matrices, classification reports, test accuracy, and false positive rates, it was observed that SVM achieved the best overall performance. SVM obtained the highest test accuracy of approximately 87% and the lowest false positive rate of about 9%, indicating that it produces fewer false alarms compared to the other models. In a forest monitoring system, minimizing false positives is important to avoid unnecessary alerts for normal wildlife sounds. Therefore, SVM is recommended as the final model for detecting threat sounds in the dataset due to its higher accuracy and better reliability.